server/coordinator.go

```diff
@@ -103,17 +105,32 @@ func (c *coordinator) dispatch(region *core.RegionInfo) {
 		}
 	}
 
-	// Check replica operator.
-	if c.limiter.OperatorCount(schedule.OpReplica) >= c.cluster.GetReplicaScheduleLimit() {
-		return
-	}
 	// Generate Operator which moves region to targeted namespace store
 	if op := c.namespaceChecker.Check(region); op != nil {
 		c.addOperator(op)
 	}
+
+	// Check replica operator.
+	if c.limiter.OperatorCount(schedule.OpReplica) >= c.cluster.GetReplicaScheduleLimit() {
+		return
+	}
 	if op := c.replicaChecker.Check(region); op != nil {
 		c.addOperator(op)
 	}
+
+	// Check merge operator.
+	if c.limiter.OperatorCount(schedule.OpMerge) >= c.cluster.GetMergeScheduleLimit() {
+		return
+	}
+	if op1, op2 := c.mergeChecker.Check(region); op1 != nil && op2 != nil {
+		// make sure two operators can add successfully altogether
+		if old := c.getOperator(op1.RegionID()); old == nil || isHigherPriorityOperator(op1, old) {
+			if old = c.getOperator(op2.RegionID()); old == nil || isHigherPriorityOperator(op2, old) {
+				c.addOperator(op1)
```

Comment:
There is still a chance that after 2 `getOperator`s returned, a new operator is inserted into the map.